### PR TITLE
Add Builder API.

### PR DESCRIPTION
**Pull request by grahamgower**
Initial sketch of a possible Builder/Factory-style API, as suggested in #208. The code here is just for discussion and has not been tested. But I think the `Builder` class is fairly complete, and we can offload most of the error checking to the `resolve()` stage. I'm luke-warm on the DemeProxy helper class. Thoughts?

### CODE REVIEW

**Comment by jeromekelleher** [comment]
LGTM. Some high-level thoughts:

1. We should provide an API for setting defaults, mirroring the spec
2. It might be worth looking at refactoring the resolution logic out of the Graph class, and into here (or some intermediate layer). Not necessary at all for a first pass, though.
3. The DemeProxy seems fine to me; we can also add a ``Builder.add_epoch(deme_id, ...)``. Both will be useful in different contexts, I think.

**Comment by grahamgower** [comment]
There's still more test coverage needed, and few cleanups are warranted after all the churn, but this should give a pretty good indication of where things are headed.

Actually, it seems that separating the Builder from the Graph isn't wholly necessary. The nicest restructuring here is enabled by switching from `add_deme(..., epochs: List[demes.Epoch] = ...)` to `add_deme(..., epochs: List[dict] = ...)`. This change alone fixes most the awkwardness pointed out in #201.

**Comment by grahamgower** [comment]
Hey @jeromekelleher, I think this is more or less done, so it would be great if you could take another look. There are a bunch of changes that went in here because it didn't make much sense to do a refactor, only to refactor again to conform with the spec. So after this PR, we should have parity with the reference implementation.

File: demes/demes.py [hunk 1]
```diff
@@ -991,50 +1030,108 @@ class Deme:
...
         for proportion in self.proportions:
             unit_interval(self, attribute, proportion)
             positive(self, attribute, proportion)
 
     @epochs.validator
     def _check_epochs(self, _attribute, _value):
         # every deme must have at least one epoch
-        if len(self.epochs) == 0:
-            raise ValueError("Demes must be defined with at least one epoch")
+        # if len(self.epochs) == 0:
```
Comment (jeromekelleher) [inline demes/demes.py:1053]: Delete?

File: demes/demes.py [hunk 2]
```diff
@@ -1937,3 +2036,196 @@ def collapse_demes(pairs):
...
+        :param float generation_time: The generation time of demes, in units given
+            by the ``time_units`` parameter. Concretely, dividing all times
+            by ``generation_time`` will convert the graph to have time
+            units in generations.  If ``generation_time`` is ``None``, the units
+            are assumed to be in generations already.
+        :param doi: If the graph describes a published demography, the DOI(s)
+            should be be given here as a list.
+        :vartype doi: list of str
+        """
+        self.data: MutableMapping[str, Any] = dict(time_units=time_units)
```
Comment (jeromekelleher) [inline demes/demes.py:2082]: Maybe add the lists here, to simplify things a bit later? E.g, ``dict(time_units=time_units, demes=[], migrations=[]...)``

**Review by jeromekelleher — APPROVED** [review]
I like it. Making the Builder responsible for working with the data model exclusively and defining its output in terms of the dict representation is a great idea. This makes the role of the class very well defined. We can make a long-term stable API on top of this.

Happy to merge whenever you are.

File: demes/demes.py (hunk 2 above)
Comment (grahamgower) [reply demes/demes.py:2079]: I had that at some point, but figured that having the explicit check and list construction inside `add_deme()` provides slightly greater robustness/flexibility in the case when someone uses `Builder.fromdict()`.

**Pull request merged**